src/v/cluster/partition.cc

```diff
@@ -322,7 +322,56 @@ ss::future<> partition::stop() {
 
 ss::future<std::optional<storage::timequery_result>>
 partition::timequery(storage::timequery_config cfg) {
-    return _raft->timequery(cfg);
+    std::optional<storage::timequery_result> result;
+    if (
+      _cloud_storage_partition && _raft->log().start_timestamp() >= cfg.time
+      && _cloud_storage_partition->is_data_available()) {
+        // We have data in the remote partition, and all the data in the raft
+        // log is ahead of the query timestamp, so proceed to query the
+        // remote partition to try and find the earliest data that has timestamp
+        // >= the query time.
+        vlog(
+          clusterlog.debug,
+          "timequery (cloud) {} t={} max_offset(k)={}",
+          _raft->ntp(),
+          cfg.time,
+          cfg.max_offset);
+
+        // remote_partition pre-translates offsets for us, so no call into
+        // the offset translator here
+        auto cloud_result = co_await _cloud_storage_partition->timequery(cfg);
```

Comment:
the `_cloud_storage_partition->is_data_available` check in the conditional above provides that remote_partition::_manifest is non-empty, but the direct pre-condition on `_cloud_storage_partition->timequery(cfg);` from looking at its implementation is that remote_partition::_segments is non-empty. do these two move in lock-step or is there some other check that provides that precondition?

Reply:
EDIT: ignore, i think. I was confusing interfaces. I meant `partition_manfiest::timequery` has that pre-condition. So this seems right.